src/action-sheets/index.js

```diff
@@ -232,6 +288,45 @@ const cancel = params => {};
 cancel.title = 'Cancel';
 cancel.errorMessage = 'Failed to hide menu';
 
+export const constructStreamActionButtons = ({
+  backgroundData: { ownUser, subscriptions },
+  streamId,
+}: {|
+  backgroundData: $ReadOnly<{
+    ownUser: User,
+    subscriptions: Map<number, Subscription>,
+    ...
+  }>,
+  streamId: number,
+|}): Button<StreamArgs>[] => {
+  const buttons = [];
+
+  const sub = subscriptions.get(streamId);
+  if (sub) {
+    if (!sub.in_home_view) {
+      buttons.push(unmuteStream);
+    } else {
+      buttons.push(muteStream);
+    }
+    if (sub.pin_to_top) {
+      buttons.push(unpinFromTop);
+    } else {
+      buttons.push(pinToTop);
+    }
+    if (sub.push_notifications === true) {
+      buttons.push(disableNotifications);
+    } else {
+      buttons.push(enableNotifications);
```

Comment:
This logic isn't quite right. See the existing place we look at this `push_notifications` property, in `StreamSettingsScreen`:
```js
    const currentValue = subscription?.push_notifications ?? userSettingStreamNotification;
```

That is, if `push_notifications` is nullish, then that doesn't necessarily mean the same as false -- it means it falls back to the value the user has chosen as their default, which may be true.

We should probably put that logic in one place; it's already a little awkward that it's duplicated within `StreamSettingsScreen.js`, and it'd be much more awkward (and much more risk of divergence) to have it duplicated across multiple files.

Reply:
deduplicated it in `StreamSettingsScreen` by creating a common function in `src/stream`; although I am not sure if it's placed at the right location.

Reply:
Looks good, thanks! See nits at https://github.com/zulip/zulip-mobile/pull/4899#discussion_r722831289 . In particular the location I'd put it is in a new file `src/subscriptions/subscriptionsModel.js`; discussion in that comment.